capsules/extra/src/hc_sr04.rs

```diff
@@ -0,0 +1,187 @@
+    }
+
+    /// Start a distance measurement.
+    fn read_distance(&self) -> Result<(), ErrorCode> {
+        if self.status.get() == Status::Idle {
+            self.status.set(Status::TriggerPulse);
+            self.trig.set();
+
+            // Setting the alarm to send the trigger pulse.
+            // According to the HC-SR04 datasheet, a 10 µs pulse should be sufficient
+            // to trigger the measurement. However, in practical tests, using this
+            // 10 µs value led to inaccurate measurements.
+            // We have chosen to use a 1 ms pulse instead because it provides stable
+            // operation and accurate measurements, even though it is slightly longer
+            // than the datasheet recommendation. While this adds a small delay to the
+            // triggering process, it does not significantly affect the overall performance
+            // of the sensor.
+            self.alarm
+                .set_alarm(self.alarm.now(), self.alarm.ticks_from_ms(1));
+            Ok(())
+        } else {
+            Err(ErrorCode::BUSY)
+        }
+    }
+
+    /// Get the maximum distance the sensor can measure in mm
+    fn get_maximum_distance(&self) -> u32 {
+        // The maximum distance is determined by the maximum pulse width the sensor can detect.
+        // As specified in the datasheet: https://www.handsontec.com/dataspecs/HC-SR04-Ultrasonic.pdf,
+        // the maximum measurable distance is approximately 4 meters.
+        // Convert this to millimeters.
+        4000
+    }
+
+    /// Get the minimum distance the sensor can measure in mm.
+    fn get_minimum_distance(&self) -> u32 {
+        // The minimum distance is determined by the minimum pulse width the sensor can detect.
+        // As specified in the datasheet: https://www.handsontec.com/dataspecs/HC-SR04-Ultrasonic.pdf,
+        // the minimum measurable distance is approximately 2 cm.
+        // Convert this to millimeters.
+        20
+    }
+}
+
+impl<'a, A: Alarm<'a>> AlarmClient for HcSr04<'a, A> {
+    /// Handle the alarm event.
+    fn alarm(&self) {
+        match self.status.get() {
+            Status::TriggerPulse => {
+                self.status.set(Status::EchoStart); // Update status to waiting for echo.
+                self.echo.enable_interrupts(InterruptEdge::RisingEdge); // Enable rising edge interrupt on echo pin.
+                self.trig.clear(); // Clear the trigger pulse.
+                self.alarm.set_alarm(
+                    self.alarm.now(),
+                    self.alarm.ticks_from_ms(MAX_ECHO_DELAY_MS),
+                ); // Set alarm for maximum echo delay.
+            }
+            // Timeout for echo pulse.
+            Status::EchoStart => {
+                self.status.set(Status::Idle); // Update status to idle.
+                if let Some(distance_client) = self.distance_client.get() {
+                    distance_client.callback(Err(ErrorCode::NOACK));
```

Comment:
This error is not documented in the HIL. `NOACK` probably makes sense but needs to be documented as a possible return value.